File: problem25/Solution.java
import java.util.Stack;

class Solution {
    public static int[] asteroidCollision(int[] asteroids) {
		Stack<Integer> stack = new Stack<>();

        for (int asteroid : asteroids) {
			if (asteroid > 0 || stack.empty() || stack.peek() < 0)
				stack.push(asteroid);
			else {
				int last = asteroid, top = stack.peek();
				while ((top > 0) && (top <= -asteroid) && (!stack.empty()) && (last != -asteroid)) {
					last = stack.pop();
					if (! stack.empty())
						top = stack.peek();
				}
				if ((stack.empty() || top < 0) && last != -asteroid)
					stack.push(asteroid);
			}
        }

        // Convert stack to array
        int[] result = new int[stack.size()];
        for (int i = result.length - 1; i >= 0; i--) {
            result[i] = stack.pop();
        }

        return result;
    }
	public static void main(String[] args) {
		int asteroids[] = {-2,-2,1,-1};
		int ans[] = asteroidCollision(asteroids);

		for (int i : ans)
			System.out.println(i);
	}
}

File: problem25/Solution.py
def asteroidCollision(asteroids):
    stack = []
    for asteroid in asteroids:
        # asteroid going right
        if asteroid > 0:
            stack.append(asteroid)
            continue
        # asteroid going left
        asteroidDestroyed = False
        while stack:
            prev = stack[-1]
            # both going left
            if prev < 0:
                stack.append(asteroid)
                break
            # collision
            absVal = abs(asteroid)
            # small asteroid --> <-- big asteroid
            if prev < absVal: stack.pop()
            else:
                # colliding asteroids are of same size
                if prev == absVal: stack.pop()
                asteroidDestroyed = True
                break
        # no asteroids in stack
        if not asteroidDestroyed and not stack:
            stack.append(asteroid)
    return stack

def main():
    asteroidsList = [[5,10,-5],[8,-8],[10,2,-5]]
    for asteroids in asteroidsList:
        print("stack = ", asteroidCollision(asteroids))

if __name__ == "__main__":
    main()
